### DISCUSSION TITLE
Observer pattern on boundary of Rust/Dart

### DISCUSSION THREAD

**Discussion by Corban-Dallas** [fzyzcjy/flutter_rust_bridge#1775] [Q&A]
Greetings!

I am stuck on the problem of implementing an event-subscription interface between Rust and Dart. What I am trying to achieve is to implement something similar to the Observer or Delegate pattern, with an event provider on the Rust side and observers on the Dart side, hidden by a trait.

It seems that it is currently impossible for the bridge to work with traits. I am certain that I should change my approach and move away from using traits, opting for something alternative but similar. The `DartOpaque` coming in mind, but for rust side it is just almost like a raw pointer without real interface, isn't it?

Due to my lack of experience with Rust, I am struggling to come up with a good idea. Perhaps someone can share some good ideas or there are already well-known practices for solving this problem?

### psuedo code of desired bridge interface behavior
```rust
// Rust side

pub fn createPublisher() -> Publisher { ... }

// Rust publisher
#[frb(opaque)]
struct Publisher {
    observer: Vec<Weak<dyn Observer>>
}

impl Publisher {
   pub fn subscribe(&self, observer: Weak<dyn Observer>) { ... }
}

// Observer interface
pub trait Observer: Sync + Send + ... {
    fn reset(&self)
    fn add(&self, items: Vec<Item>)
    ...
}
```
```dart
// Dart side

class MyObserver extends BridgeObserver { 
    void reset() { ... }
    void add(items: List<Item>) { ... }
    ...
 }

final publisher = await api.createPublisher()
final observer = MyObserver()
await publisher.subscribe(observer)

**Comment by fzyzcjy** [comment]
> It seems that it is currently impossible for the bridge to work with traits. 

Feel free to PR to implement it! It may be translated to an (abstract) class on the dart side.

>  The DartOpaque coming in mind, but for rust side it is just almost like a raw pointer without real interface, isn't it?
Due to my lack of experience with Rust, I am struggling to come up with a good idea. Perhaps someone can share some good ideas or there are already well-known practices for solving this problem?

What about using https://cjycode.com/flutter_rust_bridge/guides/direction/rust-call-dart
